File: problems/735-Asteroid-Collision/solution.go
type Stack []int

func (s *Stack) Push(num int) {
	*s = append(*s, num)
}

func (s *Stack) Pop() {
	*s = (*s)[:len(*s)-1]
}

func (s *Stack) Top() int {
	return (*s)[len(*s)-1]
}

func (s *Stack) Empty() bool {
	return len(*s) == 0
}

func asteroidCollision(asteroids []int) []int {
	stack := make(Stack, 0, len(asteroids))

	for _, val := range asteroids {
		if val > 0 {
			stack.Push(val)
		} else {
			for !stack.Empty() && stack.Top() > 0 && stack.Top() < -val {
				stack.Pop()
			}
			if stack.Empty() || stack.Top() < 0 {
				stack.Push(val)
			}
			if stack.Top() == -val {
				stack.Pop()
			}
		}
	}

	return stack
}

File: problems/735-Asteroid-Collision/solution.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        for (int asteroid : asteroids) {
            if (stack.empty() || asteroid > 0 || asteroid * stack.peek() > 0) {
                stack.push(asteroid);
                continue;
            }

            boolean flag = true;
            while (!stack.empty() && stack.peek() > 0) {
                if (Math.abs(asteroid) > Math.abs(stack.peek())) {
                    stack.pop();
                    continue;
                }

                if (Math.abs(asteroid) == Math.abs(stack.peek())) {
                    stack.pop();
                    flag = false;
                    break;
                }

                flag = false;
                break;
            }

            if (flag) {
                stack.push(asteroid);
            }
        }

        return stack.stream().mapToInt(Integer::intValue).toArray();
    }
}

